src/alire/alire-toml_index.adb

```diff
@@ -343,6 +343,48 @@ package body Alire.TOML_Index is
       Dirs.End_Search (Search);
    end Load_Catalog;
 
+   ----------------------------
+   -- Load_Release_From_File --
+   ----------------------------
+
+   function Load_Release_From_File
+     (Filename    : String;
+      Environment : Environment_Variables)
+      return        Releases.Release
+   is
+      Pkg      : Package_Type;
+      Result   : Load_Result;
+      Value    : TOML.TOML_Value;
+      Releases : Containers.Release_Sets.Set;
+   begin
+      Value := Load_TOML_From_File (Filename, Result);
+      if not Result.Success then
+         raise Constraint_Error with "could not load " & Filename;
+      end if;
+
+      --  Convert it to our intermediate data structures
+
+      Decode_TOML_Package
+        (Filename, Ada.Directories.Base_Name (Filename), Environment, Value, Pkg, Result);
+      if not Result.Success then
+         raise Constraint_Error with +Result.Message;
```

Comment:
Except when the language enforces the use of a Constraint_Error exception (for instance when decoding an integer from a string) and if practicle in general, I think it’s better to use specific types to convey errors to handle, as I did here with `Load_Result`, and leave exceptions for bona fide bugs.

The reasons for that is that 1) it’s just too easy to just forget exception handling in callers, and in general exceptions make it hard to have a good understanding of what can raise and when/how to handle errors, and 2) IIRC the language does not guarantee that it will preserve the whole exception message; I think GNAT will preserve only the first 200 characters: it’s just too easy to hit this limit with user error messages, especially when they include filenames!

In this specific case, I would advise to use an out `Load_Result` parameter to communicate to callers whether this is successful.

Reply:
Note that this is a truly exceptional situation, since the file this is intended to load is generated by alr and should always be correct. That said, I'll follow suit with the rest of the package and use the Result record.

Reply:
And just remembered that users can edit the file manually so I retract my comment :s

Reply:
:-) This objective is hard to reach, but I think it’s valuable to avoid as much as possible uncaught exceptions in error cases: they make the tool look unpolished…